File: functional-programming/src/main/haskell/Ttt.hs
import Data.List
import Control.Parallel

data Board = Board {
    crosses :: [Int],
    noughts :: [Int], 
    alternated :: Bool
} deriving Show

boardSize :: Int
boardSize = 3

ticsForWin :: Int
ticsForWin = 3

emptySlots :: Board -> [Int]
emptySlots board = filter (\position -> notElem position (noughts board ++ crosses board)) 
    $ [1  .. (boardSize * boardSize)]

addTic :: Board -> Int -> Board
addTic board cross = (Board (noughts board) (cross:(crosses board)) (not $ alternated board))

isLeftBorder :: Int -> Bool
isLeftBorder pos = if ( ((pos - 1) `mod` boardSize) == 0)
    then True
    else False

isRightBorder :: Int -> Bool
isRightBorder pos = isLeftBorder $ pos + 1

isBorder :: Int -> Bool
isBorder pos = isLeftBorder pos || isRightBorder pos

hasCross ::  Board -> Int -> Bool
hasCross board position =  position `elem` (crosses board)

from :: Board -> Int -> ( Int -> Int ) -> [Int]
from board pos move = if (isValid pos move && hasCross board (move pos) )
    then (move pos):(from board (move pos) move)
    else []

isValid :: Int -> (Int -> Int) -> Bool
isValid pos move = not (isBorder pos && isBorder (move pos))

horizontal =   ( \pos -> pos - 1            , \pos -> pos + 1 )
diagonalUp =   ( \pos -> pos - boardSize + 1, \pos -> pos + boardSize - 1)
diagonalDown = ( \pos -> pos - boardSize - 1, \pos -> pos + boardSize + 1)
vertical =     ( \pos -> pos - boardSize    , \pos -> pos + boardSize    )

isFull :: Board -> Bool
isFull board = emptySlots board == []

rows :: Board -> Int -> [[Int]]
rows board index = filter (\row -> length row > 1) $ 
    [from board index start ++ [index] ++ from board index end
    | (start, end) <- [ horizontal
                      , diagonalUp
                      , diagonalDown
                      , vertical
                      ]
    ]

hasWinningRow :: Board -> [Int] -> Bool
hasWinningRow board [] = False
hasWinningRow board (x:xs) = if (any (\row -> length row == ticsForWin) (rows board x))
    then True
    else hasWinningRow board xs

score board = let hasWin = hasWinningRow (alternate board) (crosses (alternate board)) in
    if (hasWin)
    then if (alternated board)
        then 1
        else - 1
    else 0
    where
        alternate = \board -> Board (noughts board) (crosses board) (not (alternated board))

data Tree a = Leaf a | Branch [Tree a] deriving Show

gameTree :: Board -> Tree Board
gameTree board = 
    if (isFull board || abs (score board) == 1)
    then Leaf board
    else Branch $ map (\emptySlot -> gameTree (addTic board emptySlot)) $ emptySlots board

parMap :: (a -> b) -> [a] -> [b]
parMap f [] = []
parMap f (x:xs) = fx `par` fxs `seq` (fx:fxs)
	where
		fx = f x
		fxs = parMap f xs

maximize :: Tree Board -> Int
maximize (Leaf board) = score board
maximize (Branch subGames) = maximum (parMap minimize subGames)

minimize :: Tree Board -> Int
minimize (Leaf board) = score board
minimize (Branch subGames) = minimum (parMap maximize subGames)

main = do 
	putStrLn $ show $ maximize $ gameTree $ Board [] [] False

File: functional-programming/src/main/scala/Ttt.scala
import scala.collection.immutable.BitSet

class Board(val crosses: BitSet = BitSet(), val noughts: BitSet = BitSet(), val alternated: Boolean = false, val boardSize: Int = 3) {

  val fullSet = (BitSet(1) /: (2 to boardSize * boardSize))(_ ^ BitSet(_))

  def emptySlots: BitSet = crosses ^ noughts ^ fullSet

  def addTic(tic: Int): Board =
    if ((crosses contains tic) || (noughts contains tic))
      throw new IllegalArgumentException("Position already taken")
    else
      new Board(noughts, crosses ^ BitSet(tic), !alternated)

  def isLeftBorder(pos: Int): Boolean = (pos - 1) % boardSize == 0
  def isRightBorder(pos: Int): Boolean = isLeftBorder(pos + 1)

  def isValidConnection(pos: Int, move: (Int => Int)): Boolean =
    !((isLeftBorder(pos) && isRightBorder(move(pos)))
      || (isRightBorder(pos) && isLeftBorder(move(pos))))

  def completeRowToDirection(pos: Int, direction: (Int => Int)): BitSet =
    if (isValidConnection(pos, direction) && crosses.contains(direction(pos)))
      BitSet(direction(pos)) ^ completeRowToDirection(direction(pos), direction)
    else
      BitSet()

  def rows(pos: Int): List[BitSet] = {
    if (!(crosses contains pos)) List()
    val horizontal = ((pos: Int) => pos + 1, (pos: Int) => pos - 1)
    val vertical = ((pos: Int) => pos + boardSize, (pos: Int) => pos - boardSize)
    val diagonalUp = ((pos: Int) => pos - boardSize + 1, (pos: Int) => pos + boardSize - 1)
    val diagonalDown = ((pos: Int) => pos - boardSize - 1, (pos: Int) => pos + boardSize + 1)
    for (direction <- List(horizontal, diagonalUp, diagonalDown, vertical))
      yield completeRowToDirection(pos, direction _1) ^ BitSet(pos) ^ completeRowToDirection(pos, direction _2)
  }

  def hasWinningRow: Boolean = !crosses.forall(rows(_).forall(_.size < boardSize))

  def score: Int = {
    val alternatedBoard = new Board(noughts, crosses, !alternated)
    if (alternatedBoard hasWinningRow) 
      if (alternated) 1 else -1
    else 0
  }
}

sealed trait Tree[A];
case class Leaf[A](a: A) extends Tree[A];
case class Branch[A](ts: Stream[Tree[A]]) extends Tree[A]

object Game {

  def gameTree(board: Board): Tree[Board] =
    if (board.emptySlots.size == 0 || board.score != 0)
      Leaf(board)
    else
      Branch(board.emptySlots.toStream.map((pos: Int) => gameTree(board addTic pos)))

  def maximize(tree: Tree[Board]): Int = {
    tree match {
      case Leaf(board) => board.score
      case Branch(subgames) => subgames.map((tree: Tree[Board]) => minimize(tree)).max
    }
  }

  def minimize(tree: Tree[Board]): Int = {
    tree match {
      case Leaf(board) => board.score
      case Branch(subgames) => subgames.map((tree: Tree[Board]) => maximize(tree)).min
    }
  }

}

object TicTacToe extends App {
  import Game._
  val start = System.currentTimeMillis
  val gtree = gameTree(new Board)
  val gtreeCreated = System.currentTimeMillis

  if (maximize( gtree ) == 0) println ("Game is a draw.")
  val stop = System.currentTimeMillis
  
  val sec = (t:Long) => ( (t - start) / 1000.0)
  println("Time for gametree creation: " + sec(gtreeCreated) + "s and for finding the solution " + sec(stop) + "s.")
}